refiner-lib/src/tx_hash_tracker.rs

```diff
@@ -0,0 +1,257 @@
+        }
+
+        let rx_iter = near_block
+            .shards
+            .iter()
+            .flat_map(|s| s.receipt_execution_outcomes.iter());
+
+        // Track receipts created from other receipts
+        for rx in rx_iter {
+            let rx_hash = &rx.receipt.receipt_id;
+            let tx_hash = match self.get_tx_hash(rx_hash) {
+                Some(tx_hash) => tx_hash,
+                None => {
+                    tracing::warn!("Transaction provenance unknown for receipt {}", rx_hash);
+                    continue;
+                }
+            };
+            for rx_hash in rx.execution_outcome.outcome.receipt_ids.iter() {
+                self.inner.record_rx(*rx_hash, tx_hash, block_height)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn on_block_end(&mut self, block_height: u64) -> anyhow::Result<()> {
+        self.inner.prune_state(block_height)
+    }
+}
+
+/// This struct is intentionally private and contains the core state-management logic of the
+/// transaction hash tracker. The reason to separate this from the public struct above is to
+/// allow the implementation details of the state to change without impacting the public
+/// interface.
+/// The cache in the implementation allows for fast (in-memory) look-ups, while the rocksdb
+/// storage layer enables crash recovery without data loss.
+struct TxHashTrackerImpl {
+    cache: lru::LruCache<CryptoHash, CryptoHash>,
+    persistent_storage: rocksdb::DB,
+}
+
+/// At 64 bytes per entry (two 32-byte hashes), this caps the memory footprint of the tracker
+/// at around 70 MB, which seems reasonable. One million entries should also be sufficient to
+/// ensure the cache never miss under normal conditions; it would require a receipt to be
+/// created and then not included in a block before one million other receipts we created first.
+/// With the maximum daily number of transactions ever observed on NEAR at just over two million,
+/// this means the receipt would not have been included in a block for at least 8 hours; an
+/// extremely unlikely event (typically receipts are included in the next block after they are
+/// created, less than 2 seconds later).
+const CACHE_SIZE: usize = 1_000_000;
+
+/// This is the number of block heights into the past the DB will remember receipt hashes.
+/// With a 1 second block time, this corresponds to 5 days of transactions, or 10 "epochs"
+/// since NEAR epochs tend to be around 12 hours. Keeping data for 5 epochs is the standard for
+/// non-archival nearcore nodes, so 10 epochs should be more than enough for us. At NEAR's peak of
+/// two million transactions per day, and assuming each transaction has 5 receipts on average
+/// (likely an overestimate), then this will mean fifty million entries in the DB at most.
+/// With 72 bytes per entry (two 32-byte hashes plus one 8-byte height), this will cap the
+/// storage used by this DB at under 4 GB.
+const PERSISTENT_HISTORY_SIZE: u64 = 432_000;
```

Comment:
@JonathanLogan if you could take a look at these paragraphs (re: reasonable limits), I would appreciate it.

Reply:
I think those limits are very conservative. In any case they dont create undue load on the infrastructure of our partners or us, so we can afford to be conservative.
The main thing to keep in mind here is that we also have to make sure that the refiner maintains that cache after restart, or alternatively rebuilds it.
In the latter case the assumptions might be too conservative, since we would not want to reindex the last few epochs.
Personally I would thus opt for slightly less conservative assumptions:
- track one epoch at most (I have the sense that you cannot actually have enough gas for a transaction to run for a whole epoch)

Reply:
For more context, the approx. 70 MB cache is in-memory and is less than 1 epoch in size (that 1 million number is how many receipts it remembers, and there have been more than 1 million receipts in past epochs). The persistent history is stored on-disk (in rocksdb) and it is only used to rebuilt the in-memory cache on start-up. Populating that cache should be fast since it means reading just 70 MB of data from disk, so it shouldn't add that much overhead to start up. The size of the persistent storage essentially determines how far back the refiner can be restarted without missing the transaction hash for any receipt. I don't know how far back restarts can be that aren't simply full reindexing, but I thought 10 epochs would be plenty (eg if a problem is noticed in the refiner after a Near protocol upgrade and we need to redo a whole epoch, that is possible).

> you cannot actually have enough gas for a transaction to run for a whole epoch

Yes, this is true, but what I was worried about was what happens in the case of congestion. If blocks are full then receipts can be delayed multiple blocks until there is space for them. In this case the transaction isn't spending gas to move the receipt from one block to the next, it's just the network delaying it because there's too much going on. Holding 1 million receipts is still conservative, but it's also important that this cache never miss because we do not fall back on the persistent storage in that case. The reason is because the persistent storage is optimized for reads/writes over block height ranges (this enables faster cache loading at start-up and faster pruning away old data) as opposed to random access of a particular receipt.

But long story short, if these limits are acceptable then I think it's better to be more conservative than not.